azure-ipam/ipam_test.go

```diff
@@ -0,0 +1,303 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/Azure/azure-container-networking/cns"
+	cniSkel "github.com/containernetworking/cni/pkg/skel"
+	cniTypes "github.com/containernetworking/cni/pkg/types"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+type MockCNSClient struct{}
+
+var errFoo = errors.New("err")
+
+func (c *MockCNSClient) RequestIPAddress(ctx context.Context, ipconfig cns.IPConfigRequest) (*cns.IPConfigResponse, error) {
+	switch ipconfig.InfraContainerID {
+	case "failRequestCNSArgs":
+		return nil, errFoo
+	case "failProcessCNSResp":
+		result := &cns.IPConfigResponse{
+			PodIpInfo: cns.PodIpInfo{
+				PodIPConfig: cns.IPSubnet{
+					IPAddress:    "10.0.1.10.2",
+					PrefixLength: 24,
+				},
+				NetworkContainerPrimaryIPConfig: cns.IPConfiguration{
+					IPSubnet: cns.IPSubnet{
+						IPAddress:    "10.0.1.0",
+						PrefixLength: 24,
+					},
+					DNSServers:       nil,
+					GatewayIPAddress: "10.0.0.1",
+				},
+				HostPrimaryIPInfo: cns.HostIPInfo{
+					Gateway:   "10.0.0.1",
+					PrimaryIP: "10.0.0.1",
+					Subnet:    "10.0.0.0/24",
+				},
+			},
+			Response: cns.Response{
+				ReturnCode: 0,
+				Message:    "",
+			},
+		}
+		return result, nil
+	default:
+		result := &cns.IPConfigResponse{
+			PodIpInfo: cns.PodIpInfo{
+				PodIPConfig: cns.IPSubnet{
+					IPAddress:    "10.0.1.10",
+					PrefixLength: 24,
+				},
+				NetworkContainerPrimaryIPConfig: cns.IPConfiguration{
+					IPSubnet: cns.IPSubnet{
+						IPAddress:    "10.0.1.0",
+						PrefixLength: 24,
+					},
+					DNSServers:       nil,
+					GatewayIPAddress: "10.0.0.1",
+				},
+				HostPrimaryIPInfo: cns.HostIPInfo{
+					Gateway:   "10.0.0.1",
+					PrimaryIP: "10.0.0.1",
+					Subnet:    "10.0.0.0/24",
+				},
+			},
+			Response: cns.Response{
+				ReturnCode: 0,
+				Message:    "",
+			},
+		}
+		return result, nil
+	}
+}
+
+func (c *MockCNSClient) ReleaseIPAddress(ctx context.Context, ipconfig cns.IPConfigRequest) error {
+	switch ipconfig.InfraContainerID {
+	case "failRequestCNSReleaseIPArgs":
+		return errFoo
+	default:
+		return nil
+	}
+}
+
+const (
+	happyPodArgs    = "K8S_POD_NAMESPACE=testns;K8S_POD_NAME=testname;K8S_POD_INFRA_CONTAINER_ID=testid"
+	nothappyPodArgs = "K8S_POD_NAMESPACE=testns;K8S_POD_NAME=testname;K8S_POD_INFRA_CONTAINER_ID=testid;break=break"
+)
+
+type scenario struct {
+	name    string
+	args    *cniSkel.CmdArgs
+	wantErr bool
+}
+
+func TestCmdAdd(t *testing.T) {
+	req := require.New(t)
```

Comment:
this is a little confusing (initially thought "req" was an HTTP request, not "require"). Any reason not to just call `require.NoError()` and `require.Error()` directly? (that's usually how I see this library used)